solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,275 @@
+            (address)
+        );
+        require(receiver != address(0), "0x0 receiver not allowed");
+
+        // We send wormhole tBTC OR mint canonical tBTC. We do not want to send
+        // dust. Sending wormhole tBTC is an exceptional situation and we want
+        // to keep it simple.
+        if (mintedAmount + amount > mintingLimit) {
+            bridgeToken.safeTransfer(receiver, amount);
+        } else {
+            // Function is nonReentrant and we need to update the balance with
+            // a call to completeTransferWithPayload first to know the amount.
+            // slither-disable-next-line reentrancy-benign
+            mintedAmount += amount;
+            tbtc.mint(receiver, amount);
+        }
+
+        // Function is nonReentrant and we need to extract the payload with
+        // a call to completeTransferWithPayload first.
+        // slither-disable-next-line reentrancy-events
+        emit WormholeTbtcReceived(receiver, amount);
+    }
+
+    /// @notice This function is called when the user sends their token from L2.
+    ///         The contract burns the canonical tBTC from the user and sends
+    ///         wormhole tBTC representation over the bridge.
+    ///         Keep in mind that when multiple bridges receive a minting
+    ///         authority on the canonical tBTC, this function may not be able
+    ///         to send all amounts of tBTC through the Wormhole bridge. The
+    ///         capability of Wormhole Bridge to send tBTC from the chain is
+    ///         limited to the amount of tBTC bridged through Wormhole to that
+    ///         chain.
+    /// @dev Requirements:
+    ///      - The sender must have at least `amount` of the canonical tBTC and
+    ///        it has to be approved for L2WormholeGateway.
+    ///      - The L2WormholeGateway must have at least `amount` of the wormhole
+    ///        tBTC.
+    /// @param amount The amount of tBTC to be sent.
+    /// @param recipientChain The Wormhole recipient chain ID.
+    /// @param recipient The address of the recipient in the Wormhole format.
+    /// @param arbiterFee The Wormhole arbiter fee.
+    /// @param nonce The Wormhole nonce used to batch messages together.
+    /// @return The Wormhole sequence number.
+    function sendWormhole(
+        uint256 amount,
+        uint16 recipientChain,
+        bytes32 recipient,
+        uint256 arbiterFee,
+        uint32 nonce
+    ) external nonReentrant returns (uint64) {
+        require(
+            bridgeToken.balanceOf(address(this)) >= amount,
+            "Not enough liquidity in wormhole to bridge"
+        );
+
+        emit WormholeTbtcSent(
+            amount,
+            recipientChain,
+            recipient,
+            arbiterFee,
+            nonce
+        );
+
+        mintedAmount -= amount;
+        tbtc.burnFrom(msg.sender, amount);
+
+        bridgeToken.safeApprove(address(bridge), amount);
+        return
+            bridge.transferTokens(
```

Comment:
Did you mean to `transferTokensWithPayload` here? In `receiveWormhole` you are expecting a specific payload.

If so, you will not need `arbiterFee` and will have to encode the same payload as you expect to decode in `receiveWormhole`.

Reply:
Based on the comment from @a5-pickle, if the intention is to use `transferTokensWithPayload` the contract should check if the recipient address is a trusted contract.

Reply:
Correct me if I am wrong but I think we may need another function or additional logic inside of `sendWormhole`.

When the user bridges from L1 to L2, they set `L2WormholeGateway` contract as the recipient, and they encode the L2 receiver address in the payload. On L2, the user calls `receiveWormhole` which `completeTransferWithPayload`, and the receiver address from the payload receives the canonical tBTC.

When the user bridges back from L2 to L1, they use `sendWormhole` and `transferTokens`. They set their address on L1 as the recipient. `wormholeTBTC` goes back to the bridge, and on L1 user can unlock their canonical tBTC into the recipient address straight from the Wormhole Portal Bridge contract.

The problem I believe you are talking about is when the user bridges from L2 to L2 and we should use `transferTokensWithPayload` in this case. This functionality is currently missing.

We could have two functions: 
- `sendWormhole` for cases when the receiver is an end-address (bridge back to L1).
- `sendWormholeWithPayload` for cases when the receiver is a contract (another L2).

Does it make sense? I may be missing something from the Wormhole design.

Reply:
Please see https://github.com/keep-network/tbtc-v2/pull/570/commits/5ecf2219d6fdfcb58d43df81e2e2822854d935fe and https://github.com/keep-network/tbtc-v2/pull/570/commits/ee8e218f9e03786f92bef6ccb50279d285812fe1.

The short version for those not up-to-date with the recent discussion: we had a missing path in `L2WormholeGateway` where tokens were moved between L2s and the target L2 had Wormhole tBTC gateway deployed capable of minting canonical tBTC. In this case, the transfer should be done with a payload allowing the user to receive tBTC on the target chain using the gateway deployed there.

The previous approach was fine when tBTC was transferred back to L1. It was not enough when tokens were transferred between L2s having canonical token minted by gateways.

The missing path has been covered in the `sendTbtc` function. Depending on whether there is a gateway on the target chain, transfer with payload or transfer with no payload is used.